File: Coding-Challenges/binarySearch/binary_search.py
#!/usr/bin/env python3

class Node:

    def __init__(self, key):
        self.key = key
        self.left = None
        self.right = None


class BinarySearchTree:

    def __init__(self):
        self.root = None

    def find_closest_value(self, value):
        # TODO: Your code goes here
        pass

    def pre_order(self, lst, node):
        if node is not None:
            lst.append(node.key)
            self.pre_order(lst, node.left)
            self.pre_order(lst, node.right)

    def in_order(self, lst, node):
        if node is not None:
            self.in_order(lst, node.left)
            lst.append(node.key)
            self.in_order(lst, node.right)

    def post_order(self, lst, node):
        if node is not None:
            self.post_order(lst, node.left)
            self.post_order(lst, node.right)
            lst.append(node.key)

    def insert(self, key):
        if self.root is None:
            self.root = Node(key)
        else:
            self.insert_node(self.root, key)

    def insert_node(self, node, key):
        if node.key > key:
            if node.left is None:
                node.left = Node(key)
            else:
                self.insert_node(node.left, key)
        if node.key < key:
            if node.right is None:
                node.right = Node(key)
            else:
                self.insert_node(node.right, key)

File: Coding-Challenges/binarySearch/binarySearch.js
class Node {
  constructor(key) {
    this.key = key;
    this.left = null;
    this.right = null;
  }
}

class BinarySearchTree {
  constructor() {
    this.root = null;
  }

  findClosestValue(value) {
    // TODO: Your code goes here
  }

  preOrder(arr, node) {
    if (node) {
      arr.push(node.key);
      this.preOrder(arr, node.left);
      this.preOrder(arr, node.right);
    }
  }

  inOrder(arr, node) {
    if (node) {
      this.inOrder(arr, node.left);
      arr.push(node.key);
      this.inOrder(arr, node.right);
    }
  }

  postOrder(arr, node) {
    if (node) {
      this.postOrder(arr, node.left);
      this.postOrder(arr, node.right);
      arr.push(node.key);
    }
  }

  insert(key) {
    if (this.root === null) {
      this.root = new Node(key);
    } else {
      this.insertNode(this.root, key);
    }
  }

  insertNode(node, key) {
    if (node.key > key) {
      if (node.left === null) {
        node.left = new Node(key);
      } else {
        this.insertNode(node.left, key);
      }
    }
    if (node.key < key) {
      if (node.right === null) {
        node.right = new Node(key);
      } else {
        this.insertNode(node.right, key);
      }
    }
  }
}
